src/datasets/arrow_dataset.py

```diff
@@ -3585,7 +3588,7 @@ def shuffle(
             generator = np.random.default_rng(seed)
 
         # Check if we've already cached this computation (indexed by a hash)
-        if self.cache_files:
+        if self.cache_files and not keep_in_memory:
```

Comment:
This line should stay the same:

```suggestion
        if self.cache_files:
```

Instead, [this](https://github.com/huggingface/datasets/blob/7d51531c58127d44565d6d1a6a7aa75e55d2b036/src/datasets/arrow_dataset.py#L3603) line should be replaced with:
```python
indices_cache_file_name=indices_cache_file_name if not keep_in_memory else None
```